File: chat-client/src/main/java/com/tempchat/client/service/ApiClient.java
package com.tempchat.client.service;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import okhttp3.*;

import java.io.IOException;
import java.util.ArrayList;
import java.util.List;

public class ApiClient {

    private static final MediaType JSON = MediaType.get("application/json");

    private final String baseUrl;
    private final OkHttpClient http;
    private final ObjectMapper mapper;

    public ApiClient(String baseUrl) {
        this.baseUrl = baseUrl.endsWith("/") ? baseUrl.substring(0, baseUrl.length() - 1) : baseUrl;
        this.http = new OkHttpClient();
        this.mapper = new ObjectMapper().findAndRegisterModules();
    }

    public record RoomInfo(String code, String name, String expiresAt) {}
    public record ChatMessage(String sender, String content, String sentAt) {}

    public RoomInfo createRoom(String name) throws IOException {
        String body = mapper.writeValueAsString(java.util.Map.of("name", name));
        Request req = new Request.Builder()
                .url(baseUrl + "/api/rooms")
                .post(RequestBody.create(body, JSON))
                .build();
        try (Response res = http.newCall(req).execute()) {
            assertSuccess(res);
            JsonNode node = mapper.readTree(res.body().string());
            return new RoomInfo(
                    node.get("code").asText(),
                    node.get("name").asText(),
                    node.get("expiresAt").asText());
        }
    }

    public RoomInfo getRoom(String code) throws IOException {
        Request req = new Request.Builder()
                .url(baseUrl + "/api/rooms/" + code)
                .get()
                .build();
        try (Response res = http.newCall(req).execute()) {
            assertSuccess(res);
            JsonNode node = mapper.readTree(res.body().string());
            return new RoomInfo(
                    node.get("code").asText(),
                    node.get("name").asText(),
                    node.get("expiresAt").asText());
        }
    }

    public List<ChatMessage> getMessages(String roomCode) throws IOException {
        Request req = new Request.Builder()
                .url(baseUrl + "/api/rooms/" + roomCode + "/messages")
                .get()
                .build();
        try (Response res = http.newCall(req).execute()) {
            assertSuccess(res);
            List<ChatMessage> messages = new ArrayList<>();
            for (JsonNode node : mapper.readTree(res.body().string())) {
                messages.add(new ChatMessage(
                        node.get("sender").asText(),
                        node.get("content").asText(),
                        node.get("sentAt").asText()));
            }
            return messages;
        }
    }

    private void assertSuccess(Response res) throws IOException {
        if (!res.isSuccessful()) {
            throw new IOException("Server error " + res.code() + ": " + res.message());
        }
    }
}

File: chat-client-c/Services/ApiClient.cs
using System.Net.Http.Json;
using System.Text.Json;

namespace TempChat.Services;

public sealed class ApiClient
{
    public record RoomInfo(string Code, string Name, string ExpiresAt);
    public record ChatMessage(string Sender, string Content, string SentAt);

    private static readonly HttpClient Http = new();
    private readonly string _baseUrl;

    public ApiClient(string baseUrl)
    {
        _baseUrl = baseUrl.TrimEnd('/');
    }

    public async Task<RoomInfo> CreateRoomAsync(string name)
    {
        var resp = await Http.PostAsJsonAsync($"{_baseUrl}/api/rooms", new { name });
        resp.EnsureSuccessStatusCode();
        var json = await resp.Content.ReadFromJsonAsync<JsonElement>();
        return ParseRoom(json);
    }

    public async Task<RoomInfo> GetRoomAsync(string code)
    {
        var resp = await Http.GetAsync($"{_baseUrl}/api/rooms/{code}");
        resp.EnsureSuccessStatusCode();
        var json = await resp.Content.ReadFromJsonAsync<JsonElement>();
        return ParseRoom(json);
    }

    public async Task<List<ChatMessage>> GetMessagesAsync(string roomCode)
    {
        var resp = await Http.GetAsync($"{_baseUrl}/api/rooms/{roomCode}/messages");
        resp.EnsureSuccessStatusCode();
        var json = await resp.Content.ReadFromJsonAsync<JsonElement>();
        var result = new List<ChatMessage>();
        foreach (var item in json.EnumerateArray())
            result.Add(new ChatMessage(
                item.GetProperty("sender").GetString()!,
                item.GetProperty("content").GetString()!,
                item.GetProperty("sentAt").GetString()!));
        return result;
    }

    private static RoomInfo ParseRoom(JsonElement json) =>
        new(json.GetProperty("code").GetString()!,
            json.GetProperty("name").GetString()!,
            json.GetProperty("expiresAt").GetString()!);
}
